app/src/main/java/org/wikipedia/edit/templates/TemplatesSearchActivity.kt

```diff
@@ -0,0 +1,220 @@
+import androidx.appcompat.widget.SearchView
+import androidx.core.view.isVisible
+import androidx.lifecycle.Lifecycle
+import androidx.lifecycle.lifecycleScope
+import androidx.lifecycle.repeatOnLifecycle
+import androidx.paging.LoadState
+import androidx.paging.PagingDataAdapter
+import androidx.recyclerview.widget.DiffUtil
+import androidx.recyclerview.widget.LinearLayoutManager
+import androidx.recyclerview.widget.RecyclerView
+import kotlinx.coroutines.flow.collectLatest
+import kotlinx.coroutines.launch
+import org.wikipedia.Constants
+import org.wikipedia.R
+import org.wikipedia.activity.BaseActivity
+import org.wikipedia.databinding.ActivityTemplatesSearchBinding
+import org.wikipedia.databinding.ItemTemplatesSearchBinding
+import org.wikipedia.dataclient.WikiSite
+import org.wikipedia.dataclient.mwapi.TemplateDataResponse
+import org.wikipedia.page.PageTitle
+import org.wikipedia.settings.Prefs
+import org.wikipedia.util.DeviceUtil
+import org.wikipedia.util.DimenUtil
+import org.wikipedia.util.FeedbackUtil
+import org.wikipedia.util.ResourceUtil
+import org.wikipedia.util.StringUtil
+
+class TemplatesSearchActivity : BaseActivity() {
+    private lateinit var binding: ActivityTemplatesSearchBinding
+    private lateinit var insertTemplateFragment: InsertTemplateFragment
+
+    private var templatesSearchAdapter: TemplatesSearchAdapter? = null
+
+    val viewModel: TemplatesSearchViewModel by viewModels { TemplatesSearchViewModel.Factory(intent.extras!!) }
+
+    private val searchCloseListener = SearchView.OnCloseListener {
+        closeSearch()
+        false
+    }
+
+    private val searchQueryListener = object : SearchView.OnQueryTextListener {
+        override fun onQueryTextSubmit(queryText: String): Boolean {
+            DeviceUtil.hideSoftKeyboard(this@TemplatesSearchActivity)
+            return true
+        }
+
+        override fun onQueryTextChange(queryText: String): Boolean {
+            binding.searchCabView.setCloseButtonVisibility(queryText)
+            startSearch(queryText.trim())
+            return true
+        }
+    }
+
+    public override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        binding = ActivityTemplatesSearchBinding.inflate(layoutInflater)
+        setContentView(binding.root)
+        setSupportActionBar(binding.toolbar)
+
+        initSearchView()
+
+        templatesSearchAdapter = TemplatesSearchAdapter()
+        binding.templateRecyclerView.layoutManager = LinearLayoutManager(this)
+        binding.templateRecyclerView.adapter = templatesSearchAdapter
+
+        insertTemplateFragment = supportFragmentManager.findFragmentById(R.id.insertTemplateFragment) as InsertTemplateFragment
+
+        binding.insertTemplateButton.setOnClickListener {
+            viewModel.selectedPageTitle?.let {
+                val list = Prefs.recentUsedTemplates.toMutableList()
+                list.add(it)
+                Prefs.recentUsedTemplates = list
```

Comment:
Should we also set a maximum number of items in the list, e.g. 100, so that it can't grow without bounds?

Reply:
Done.